[mlir] Mark `isa/dyn_cast/cast/...` member functions deprecated.

chsigg opened the pull request:
This also removes the member overload in TypeSwitch.

All other users have been removed in fac349a169976f822fb27f03e623fa0d28aec1f3 and bd9fdce69b4c4cdb572e715c5f453aaf9b77b83a.

chsigg committed:
[mlir] Mark `isa/dyn_cast/cast/...` member functions deprecated.

All users have been removed in fac349a169976f822fb27f03e623fa0d28aec1f3 and bd9fdce69b4c4cdb572e715c5f453aaf9b77b83a.

Except for TypeSwitch, where the deprecation warning needs to be disabled until the functions are removed.

chsigg commented:
Note that this change previously broke a [Windows build](https://lab.llvm.org/buildbot/#/builders/216/builds/38157).

@dyung, could you please advise how to test this change so I don't break it again?

dyung commented:
> Note that this change previously broke a [Windows build](https://lab.llvm.org/buildbot/#/builders/216/builds/38157).
> 
> @dyung, could you please advise how to test this change so I don't break it again?

I don't know if this change is any different from what you submitted, but when I applied the patches locally, our Windows build failed the same way as before.

chsigg committed:
Also deprecate mlir::pdll::ast::Type cast members.

Remove member cast overload in TypeSwitch so we don't have to disable the warning.

chsigg commented:
This passes the Windows CI now which uses the same Visual Studio 2019 version as the CI build that previously broke. This is expected because it no longer uses the `LLVM_SUPPRESS_DEPRECATED_DECLARATIONS_PUSH` macro. I did see cl.exe running out of heap space on previous revisions building flang, but that is likely an unrelated transient failure.

chsigg commented:
The [Windows build](https://lab.llvm.org/buildbot/#/builders/216/builds/38157) that caused problems previously is still [passing](https://lab.llvm.org/buildbot/#/builders/216/builds/38322) after this change.

DanielCChen commented:
It seems this PR causes build failure as
```
llvm-project/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:148:26: error: 'dyn_cast' is deprecated: Use mlir::dyn_cast<U>() instead [-Werror,-Wdeprecated-declarations]
    if (auto funcLoc = l.dyn_cast<mlir::FileLineColLoc>())
                         ^

```

There are also quite a lot `warning` on top of the failures. @chsigg

chsigg commented:
> It seems this PR causes build failure on LoP (Linux On Power) as
> 
> ```
> llvm-project/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:148:26: error: 'dyn_cast' is deprecated: Use mlir::dyn_cast<U>() instead [-Werror,-Wdeprecated-declarations]
>     if (auto funcLoc = l.dyn_cast<mlir::FileLineColLoc>())
>                          ^
> ```
> 
> There are also quite a lot `warning` on top of the failures. @chsigg

Sorry about that, not sure how I missed these. Hopefully they are now fixed with https://github.com/llvm/llvm-project/commit/52cb9537896992e0fa689c45469bccb36970c0c7?

Thank you very much [kazutakahirata](https://github.com/kazutakahirata) and [PeimingLiu](https://github.com/PeimingLiu) for jumping in!

DanielCChen commented:
> > It seems this PR causes build failure on LoP (Linux On Power) as
> > ```
> > llvm-project/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:148:26: error: 'dyn_cast' is deprecated: Use mlir::dyn_cast<U>() instead [-Werror,-Wdeprecated-declarations]
> >     if (auto funcLoc = l.dyn_cast<mlir::FileLineColLoc>())
> >                          ^
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There are also quite a lot `warning` on top of the failures. @chsigg
> 
> Sorry about that, not sure how I missed these. Hopefully they are now fixed with [52cb953](https://github.com/llvm/llvm-project/commit/52cb9537896992e0fa689c45469bccb36970c0c7)?
> 
> Thank you very much [kazutakahirata](https://github.com/kazutakahirata) and [PeimingLiu](https://github.com/PeimingLiu) for jumping in!

Yes. All the errors and warnings are fixed. Thanks!
